src/test/java/org/folio/services/migration/BatchedReadStreamTest.java

```diff
@@ -0,0 +1,173 @@
+  public void shouldProcessRecordsByBatch() {
+    var numberOfRecords = 410;
+    var batchSize = 100;
+    var numberOfBatches = 5;
+    var rowStream = new ReindexJobRunnerTest.TestRowStream(numberOfRecords);
+    var batchedReadStream = new BatchedReadStream<>(rowStream, batchSize);
+    var promise = Promise.promise();
+    Handler<List<Row>> handler = mock(Handler.class);
+
+    batchedReadStream
+      .handler(handler)
+      .endHandler(promise::tryComplete)
+      .exceptionHandler(promise::tryFail);
+
+    // Wait until completed
+    get(promise.future());
+
+    ArgumentCaptor<List<Row>> captor = ArgumentCaptor.forClass(List.class);
+    verify(handler, times(numberOfBatches)).handle(captor.capture());
+
+    assertThat(captor.getAllValues()).hasSize(numberOfBatches);
+    assertThat(captor.getAllValues().get(4)).hasSize(10);
+    captor.getAllValues().stream().limit(numberOfBatches - 1)
+      .forEach(list -> assertThat(list).hasSize(batchSize));
+
+
+    var ids = new HashSet<UUID>(numberOfRecords);
+    captor.getAllValues().forEach(
+      list -> list.forEach(
+        // Make sure all IDS are unique so that batch doe not produce
+        // duplicate records
+        row -> assertThat(ids.add(row.getUUID("id"))).isTrue()));
+  }
+
+  @Test
+  public void shouldDoNothingIfNullHandler() {
+    var rowStream = mockStream();
+    var batchedReadStream = new BatchedReadStream<>(rowStream);
+
+    batchedReadStream.handler(null);
+
+    verifyNoInteractions(rowStream);
+  }
+
+  @Test
+  public void shouldDelegateCallForExceptionHandler() {
+    var delegate = mockStream();
+    var batchedReadStream = new BatchedReadStream<>(delegate);
+
+    batchedReadStream.exceptionHandler(error -> {});
+
+    verify(delegate, times(1)).exceptionHandler(any());
+  }
+
+  @Test
+  public void shouldDelegateCallForEndHandler() {
+    var delegate = mockStream();
+    var batchedReadStream = new BatchedReadStream<>(delegate);
+
+    batchedReadStream.endHandler(error -> {});
+
+    verify(delegate, times(1)).endHandler(any());
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldProcessRemainingRecordsWhenEndReached() {
+    var batchSize = 10;
+    var delegate = new ReindexJobRunnerTest.TestRowStream(21);
+    var batchedReadStream = new BatchedReadStream<>(delegate, batchSize);
+    Handler<Void> endHandler = mock(Handler.class);
+    Handler<List<Row>> handler = mock(Handler.class);
+
+    batchedReadStream
+      .handler(handler)
+      .endHandler(endHandler);
+
+    await().untilAsserted(() -> {
+      var captor = ArgumentCaptor.forClass(List.class);
+      verify(endHandler, times(1)).handle(any());
+
+      verify(handler, times(3)).handle(captor.capture());
+      assertThat(captor.getAllValues()).hasSize(3);
+      assertThat(captor.getAllValues().get(0)).hasSize(batchSize);
+      assertThat(captor.getAllValues().get(1)).hasSize(batchSize);
+      assertThat(captor.getAllValues().get(2)).hasSize(1);
+    });
+  }
```

Comment:
How is this test different to `shouldProcessRecordsByBatch`?

Reply:
Removed this one